internal/worker/sshserver/server.go

```diff
@@ -126,10 +126,10 @@ func NewServerWorker(config ServerWorkerConfig) (worker.Worker, error) {
 	// Start server.
 	s.tomb.Go(func() error {
 		err := s.Server.Serve(listener)
-		if errors.Is(err, ssh.ErrServerClosed) {
-			return nil
+		if !errors.Is(err, ssh.ErrServerClosed) {
+			s.config.Logger.Errorf("serve returned an unexpected error: %v", err)
 		}
-		return errors.Trace(err)
+		return tomb.ErrDying
```

Comment:
Looking at this again, this is incorrect because of a few reasons,
1. The tomb is not necessarily dying yet, we would need to add `<-s.tomb.Dying()`.
2. Adding a `<-s.tomb.Dying()` would mean an actual error would result in the server never restarting.
3. Returning `nil` error is the same as the above, an actual error means the server will never restart.

A fix to the upstream has been proposed [here](https://github.com/gliderlabs/ssh/pull/248). I'll propose a more correct fix that doesn't rely on an upstream change as that might take a while to land.